src/main/java/com/google/flourbot/entity/EntityModuleImplementation.java

```diff
@@ -1,100 +1,99 @@
 import com.google.flourbot.datastorage.DataStorage;
 import com.google.flourbot.datastorage.FirebaseDataStorage;
-import com.google.cloud.firestore.QueryDocumentSnapshot;
-import java.util.Optional;
-import java.util.Map;
-
 import com.google.flourbot.entity.action.Action;
 import com.google.flourbot.entity.action.SheetAppendAction;
-import com.google.flourbot.entity.trigger.Trigger;
 import com.google.flourbot.entity.trigger.CommandTrigger;
+import com.google.flourbot.entity.trigger.Trigger;
 
-import java.lang.InterruptedException;
+import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
 public class EntityModuleImplementation implements EntityModule {
-    
-    private static EntityModuleImplementation instance = null;
-    private final DataStorage datastorage;
 
-    private EntityModuleImplementation() {
-        this.datastorage = new FirebaseDataStorage();
+  private static EntityModuleImplementation instance = null;
+  private final DataStorage datastorage;
+
+  private EntityModuleImplementation() {
+    this.datastorage = new FirebaseDataStorage();
+  }
+
+  public Optional<Macro> getMacro(String userEmail, String macroName)
+      throws InterruptedException, ExecutionException {
+    Optional<QueryDocumentSnapshot> optionalDocument =
+        datastorage.getDocument(userEmail, macroName);
+
+    if (!optionalDocument.isPresent()) {
+      // Handle the case if an empty document is returned.
+      return Optional.empty();
     }
 
-    public Optional<Macro> getMacro(String userEmail, String macroName) throws InterruptedException, ExecutionException {
-        Optional<QueryDocumentSnapshot> optionalDocument = datastorage.getDocument(userEmail, macroName); 
+    // Retrieve value from optional if it's not empty
+    QueryDocumentSnapshot document = optionalDocument.get();
 
-        if (!optionalDocument.isPresent()) {
-            // Handle the case if an empty document is returned.
-            return Optional.empty();
-        } 
-       
-        // Retrieve value from optional if it's not empty 
-        QueryDocumentSnapshot document = optionalDocument.get();
+    Map<String, Object> macroMap = document.getData();
 
-        Map<String, Object> macroMap = document.getData();
+    Map<String, Object> triggerMap = (Map<String, Object>) macroMap.get("trigger");
+    Map<String, Object> actionMap = (Map<String, Object>) macroMap.get("action");
 
-        Map<String, Object> triggerMap = (Map<String, Object>) macroMap.get("trigger");
-        Map<String, Object> actionMap = (Map<String, Object>) macroMap.get("action");
+    Optional<Trigger> optionalTrigger = getTrigger(triggerMap);
+    Optional<Action> optionalAction = getAction(actionMap);
 
-        Optional<Trigger> optionalTrigger = getTrigger(triggerMap);
-        Optional<Action> optionalAction = getAction(actionMap);
+    if (!optionalTrigger.isPresent() || !optionalAction.isPresent()) {
+      return Optional.empty();
```

Comment:
nice! It is an interesting case that I haven't taken into account yet - we get the macro object, but it is corrupted. I think throwing IllegalStateException would work here as well, but that's something to think about later.

Reply:
Throwing an exception might help with debugging here too.